crypto/s2n_tls13_keys.c

```diff
@@ -343,3 +344,42 @@ int s2n_tls13_update_application_traffic_secret(struct s2n_tls13_keys *keys, str
 
     return 0;
 }
+
+int s2n_tls13_derive_resumption_master_secret(struct s2n_tls13_keys *keys, struct s2n_hash_state *hashes, struct s2n_blob *secret_blob)
+{
+    notnull_check(keys);
+    notnull_check(hashes);
+    notnull_check(secret_blob);
+
+    /* Sanity check that input hash is of expected type */
+    ENSURE_POSIX(keys->hash_algorithm == hashes->alg, S2N_ERR_HASH_INVALID_ALGORITHM);
+
+    s2n_tls13_key_blob(message_digest, keys->size);
+
+    /* Copy the hashes into the message_digest */
+    DEFER_CLEANUP(struct s2n_hash_state hkdf_hash_copy, s2n_hash_free);
+    GUARD(s2n_hash_new(&hkdf_hash_copy));
+    GUARD(s2n_hash_copy(&hkdf_hash_copy, hashes));
+    GUARD(s2n_hash_digest(&hkdf_hash_copy, message_digest.data, message_digest.size));
+
+    /* Derive master session resumption from master secret */
+    GUARD(s2n_hkdf_expand_label(&keys->hmac, keys->hmac_algorithm, &keys->extract_secret,
+        &s2n_tls13_label_resumption_master_secret, &message_digest, secret_blob));
```

Comment:
NitPick: This seems to be commonly repeated code (with only the label varying). Should it be factored out? I'd also like someone to explain it to me. I trust that it is correct, but am having trouble following some of it.

Reply:
The reason why we create a hash copy is because s2n_hash_digest() calls an Openssl function which modifies the hash state. This function in Openssl is called SHAx_Final (x being the specific hash function you're using) and if you check the [documentation](https://www.openssl.org/docs/man1.1.1/man3/SHA224_Final.html) it looks like it erases the SHA_CTX variable, which in s2n terms is the hash_state. 
As for whether or not this code should be factored out, 🤷‍♀️ it's not that much code? I don't feel strongly about factoring out those specific lines.